File: nhatbot_twist_teleop/src/joy_to_twist.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joy.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"

class JoyToTwistNode : public rclcpp::Node
{
public:
    JoyToTwistNode() : Node("joy_to_twist")
    {

        this->declare_parameter("linear_scale", 0.3);
        this->declare_parameter("angular_scale", 0.3);
        this->declare_parameter("deadman_button", 9);
        this->declare_parameter("deadzone_threshold", 0.01);

        linear_scale_ = this->get_parameter("linear_scale").as_double();
        angular_scale_ = this->get_parameter("angular_scale").as_double();
        deadman_button_ = this->get_parameter("deadman_button").as_int();
        deadzone_threshold_ = this->get_parameter("deadzone_threshold").as_double();

        joy_subscriber_ = this->create_subscription<sensor_msgs::msg::Joy>(
            "/joy", 10, std::bind(&JoyToTwistNode::joy_callback, this, std::placeholders::_1));

        twist_publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>("/input_joy/cmd_vel_stamped", 10);

        RCLCPP_INFO(this->get_logger(), "✅ joy_to_twist node is running...");
    }

private:
    void joy_callback(const sensor_msgs::msg::Joy::SharedPtr joy_msg)
    {

        if (joy_msg->axes.size() <= 2 || joy_msg->buttons.size() <= deadman_button_) 
        {
            RCLCPP_WARN(this->get_logger(), "❗️Joy message does not have expected number of axes or buttons");
            return;
        }


        geometry_msgs::msg::TwistStamped twist_stamped;
        twist_stamped.header.stamp = get_clock()->now();
        twist_stamped.header.frame_id = "base_link";


        bool deadman_pressed = (joy_msg->buttons[deadman_button_] != 0);

    if (!deadman_pressed)
    {
        twist_stamped.twist.linear.x = 0.0;
        twist_stamped.twist.angular.z = 0.0;

        if (!is_zero_twist(last_twist_stamped_))  // ✅ Chỉ publish nếu lần trước chưa là 0
        {
            twist_publisher_->publish(twist_stamped);
            last_twist_stamped_ = twist_stamped;
            RCLCPP_INFO(this->get_logger(), "🛑 Deadman released → Stop robot");
        }

        return;
    }

        double linear_value = joy_msg->axes[1];  // Trục di chuyển tiến/lùi
        double angular_value = joy_msg->axes[2]; // Trục quay trái/phải

        // 🚫 Bỏ trôi nhẹ bằng deadzone
        if (std::abs(linear_value) < deadzone_threshold_)  linear_value = 0.0;
        
        if (std::abs(angular_value) < deadzone_threshold_) angular_value = 0.0;
        

        twist_stamped.twist.linear.x = linear_value * linear_scale_;
        twist_stamped.twist.angular.z = angular_value * angular_scale_;



    // ✅ Nếu đang điều khiển (≠ 0) thì LUÔN publish

    if (!is_zero_twist(twist_stamped))
    {
        twist_publisher_->publish(twist_stamped);
        last_twist_stamped_ = twist_stamped;
        first_publish_ = false;
        return;
    }


    // ✅ Nếu twist == 0, thì CHỈ publish 1 lần (khi từ trạng thái ≠ 0 trở về 0)


    if (!is_zero_twist(last_twist_stamped_))
    {
        twist_publisher_->publish(twist_stamped);
        last_twist_stamped_ = twist_stamped;
        RCLCPP_INFO(this->get_logger(), "🛑 Joystick returned to 0 → Stop robot");
    }

        // if (first_publish_ || twist_changed(twist_stamped, last_twist_stamped_))
        // {
        //     twist_publisher_->publish(twist_stamped);
        //     last_twist_stamped_ = twist_stamped;
        //     first_publish_ = false;
        // }
    }

    bool twist_changed(const geometry_msgs::msg::TwistStamped &a, const geometry_msgs::msg::TwistStamped &b)
    {
        return std::abs(a.twist.linear.x - b.twist.linear.x) > 1e-3 ||
               std::abs(a.twist.angular.z - b.twist.angular.z) > 1e-3;
    }
    bool is_zero_twist(const geometry_msgs::msg::TwistStamped &twist)
    {
        return std::abs(twist.twist.linear.x) < 1e-3 && std::abs(twist.twist.angular.z) < 1e-3;
    }

    double linear_scale_;
    double angular_scale_;
    int deadman_button_;
    double deadzone_threshold_;

    rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_subscriber_;
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr twist_publisher_;
    geometry_msgs::msg::TwistStamped last_twist_stamped_;
    bool first_publish_ = true;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<JoyToTwistNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: nhatbot_twist_teleop/nhatbot_twist_teleop/joy_to_twist.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy
from geometry_msgs.msg import Twist

class JoyToTwistNode(Node):
    def __init__(self):
        super().__init__('joy_to_twist')

        self.declare_parameter("linear_scale", 0.5)          # Hệ số tốc độ tiến/lùi
        self.declare_parameter("angular_scale", 0.5)         # Hệ số tốc độ quay
        self.declare_parameter("deadman_button", 9)          # Nút giữ để di chuyển
        self.declare_parameter("deadzone_threshold", 0.1)    # Ngưỡng tối thiểu tránh trôi
        self.declare_parameter("axis_linear_x", 1)          # Trục điều khiển tiến/lùi
        self.declare_parameter("axis_angular_z", 2)         # Trục điều khiển quay trái/phải

        # Lấy giá trị từ tham số ROS 2
        self.linear_scale = self.get_parameter("linear_scale").get_parameter_value().double_value
        self.angular_scale = self.get_parameter("angular_scale").get_parameter_value().double_value
        self.deadman_button = self.get_parameter("deadman_button").get_parameter_value().integer_value
        self.deadzone_threshold = self.get_parameter("deadzone_threshold").get_parameter_value().double_value
        self.axis_linear_x = self.get_parameter("axis_linear_x").get_parameter_value().integer_value
        self.axis_angular_z = self.get_parameter("axis_angular_z").get_parameter_value().integer_value

        # Subscribe vào topic /joy
        self.joy_subscriber = self.create_subscription(Joy, '/joy', self.joy_callback, 10)

        # Publisher gửi dữ liệu lên /teleop
        self.twist_publisher = self.create_publisher(Twist, '/nhatbot_controller/cmd_vel', 10)  # "nhatbot_controller/cmd_vel   teleop

        self.get_logger().info("✅ joy_to_twist đã khởi động với tham số từ ROS 2!")

    def joy_callback(self, joy_msg):
        """ Xử lý dữ liệu từ joystick và xuất ra Twist """
        twist = Twist()

        # Kiểm tra nếu không nhấn nút deadman -> dừng robot
        if joy_msg.buttons[self.deadman_button] == 0:
            twist.linear.x = 0.0
            twist.angular.z = 0.0
        else:
            # Đọc giá trị từ joystick
            linear_value = joy_msg.axes[self.axis_linear_x]
            angular_value = joy_msg.axes[self.axis_angular_z]

            # Nếu giá trị trục nhỏ hơn ngưỡng, set về 0 để tránh trôi nhẹ
            if abs(linear_value) < self.deadzone_threshold:
                linear_value = 0.0
            if abs(angular_value) < self.deadzone_threshold:
                angular_value = 0.0

            # Tính toán vận tốc
            twist.linear.x = linear_value * self.linear_scale
            twist.angular.z = angular_value * self.angular_scale

        # Xuất dữ liệu lên topic /teleop
        self.twist_publisher.publish(twist)

def main():
    rclpy.init()
    node = JoyToTwistNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        if rclpy.ok():
            node.get_logger().info("🛑 Ctrl+C detected! Đang dừng joy_twist_node...")
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()



